File: JWT/OTP examples - JWT/PYTHON/GenerateJWT.py
import os
import json
import time
import base64
import hashlib
import secrets
from pathlib import Path
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
import jwt
import requests

# === CONFIGURATION ===
priv_key_file = Path(__file__).with_name('private.pem')
pub_key_file = Path(__file__).with_name('public.pem')

# === USER NEEDS TO CONFIGURE THESE SETTINGS HERE ===
org_id = 165
app_workvivo = '3981'
app_aud = 'unwired.workvivo.red'
user_email = 'test@nomail'
jwt_lifetime = 600
api_url = 'https://api-gateway.workvivo.red/v1/unwired/users/otp'
app_issuer = 'org1'
app_subject = 'app'

# === KEYPAIR GENERATION OR LOADING ===

#
#  You can generate your own RSA keypair manually using OpenSSL:
#  
#  # Generate a 4096-bit private key
#  openssl genpkey -algorithm RSA -out private.pem -pkeyopt rsa_keygen_bits:4096
#  
#  # Extract the public key from the private key
#  openssl rsa -in private.pem -pubout -out public.pem
#  
#  Place both files (private.pem and public.pem) in the same directory as this script.
#  The script will use them automatically instead of generating new keys.
#  


def generate_rsa_keypair():
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=4096
    )
    priv_pem = private_key.private_bytes(
        serialization.Encoding.PEM,
        serialization.PrivateFormat.PKCS8,
        serialization.NoEncryption()
    )
    pub_pem = private_key.public_key().public_bytes(
        serialization.Encoding.PEM,
        serialization.PublicFormat.SubjectPublicKeyInfo
    )

    priv_key_file.write_bytes(priv_pem)
    pub_key_file.write_bytes(pub_pem)

    print("Generated and saved new RSA keypair.")
    return private_key, priv_pem, pub_pem

def load_keys():
    priv_pem = priv_key_file.read_bytes()
    pub_pem = pub_key_file.read_bytes()
    private_key = serialization.load_pem_private_key(priv_pem, password=None)
    print("Loaded existing RSA keypair.")
    return private_key, priv_pem, pub_pem

if not priv_key_file.exists() or not pub_key_file.exists():
    private_key, private_pem, public_pem = generate_rsa_keypair()
else:
    private_key, private_pem, public_pem = load_keys()

# === base64url encode ===
def base64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b'=').decode('utf-8')

# === BUILD JWKS ===
public_numbers = private_key.public_key().public_numbers()
e = base64url_encode(public_numbers.e.to_bytes((public_numbers.e.bit_length() + 7) // 8, byteorder='big'))
n = base64url_encode(public_numbers.n.to_bytes((public_numbers.n.bit_length() + 7) // 8, byteorder='big'))

jwk = {
    'kty': 'RSA',
    'alg': 'RS256',
    'use': 'sig',
    'n': n,
    'e': e,
}

thumbprint_input = json.dumps({'e': e, 'kty': 'RSA', 'n': n}, separators=(',', ':')).encode('utf-8')
thumbprint = base64url_encode(hashlib.sha256(thumbprint_input).digest())
jwk['kid'] = thumbprint
jwks = {'keys': [jwk]}

# === CREATE JWT ===
now = int(time.time())
payload = {
    'jti': secrets.token_hex(32),
    'iss': app_issuer,
    'sub': app_subject,
    'workvivo_id': app_workvivo,
    'aud': app_aud,
    'iat': now,
    'nbf': now,
    'exp': now + jwt_lifetime,
    'state': secrets.token_hex(32),
}

token = jwt.encode(
    payload,
    private_pem,
    algorithm='RS256',
    headers={'kid': jwk['kid']}
)

# === OUTPUT ===
print(f"\n=== JWT ===\n{token}")
print(f"\n=== kid ===\n{jwk['kid']}")
print(f"\n=== public.pem ===\n{public_pem.decode()}")
print(f"\n=== jwks ===\n{json.dumps(jwks, indent=4)}")

# prompt to confirm JWKS is uploaded or public key
input("\nPress [Enter] once you have updated your JWKS at:\nhttps://HOST/admin/developers/apps/manage\n")

# === SEND REQUEST ===
headers = {
    'Workvivo-Id': str(org_id),
    'x-workvivo-jwt': token,
    'x-workvivo-jwt-keyid': jwk['kid'],
    'Accept': 'application/json',
    'Content-Type': 'application/json',
}
post_body = json.dumps({'email': user_email})

response = requests.post(api_url, headers=headers, data=post_body, verify=False)
print(f"\n=== Response ({response.status_code}) ===\n{response.text}")

File: JWT/One-time-passcode/PHP/GenerateJWT.php
<?php

require __DIR__ . '/vendor/autoload.php';
use Firebase\JWT\JWT;

// === CONFIGURATION ===

$privKeyFile = __DIR__ . '/private.pem';
$pubKeyFile  = __DIR__ . '/public.pem';

// === USER NEEDS TO CONFIGURE THESE SETTINGS HERE ===
$orgId       = 165; // This is the organisation ID
$appWorkvivo = '3981'; // APP ID from your Workvivo app with unwiredotp.*.write permission
$appAud      = 'unwired.workvivo.red';
$userEmail   = 'test@nomail';

$jwtLifetime = 600;
$apiUrl      = 'https://api-gateway.workvivo.red/v1/unwired/users/otp'; //In case it is EU Production HOST should be api.workvivo.com
$appIssuer   = 'org1';
$appSubject  = 'app';

// === KEYPAIR GENERATION OR LOADING ===

/*
 * You can generate your own RSA keypair manually using OpenSSL:
 *
 * # Generate a 4096-bit private key
 * openssl genpkey -algorithm RSA -out private.pem -pkeyopt rsa_keygen_bits:4096
 *
 * # Extract the public key from the private key
 * openssl rsa -in private.pem -pubout -out public.pem
 *
 * Place both files (private.pem and public.pem) in the same directory as this script.
 * The script will use them automatically instead of generating new keys.
 */

if (! file_exists($privKeyFile) || ! file_exists($pubKeyFile)) {
    // generate a new 4096-bit RSA keypair
    $res = openssl_pkey_new([
        'private_key_bits' => 4096,
        'private_key_type' => OPENSSL_KEYTYPE_RSA,
    ]);
    openssl_pkey_export($res, $privatePem);
    $keyDetails   = openssl_pkey_get_details($res);
    $publicPem    = $keyDetails['key'];

    // save to file
    file_put_contents($privKeyFile, $privatePem);
    file_put_contents($pubKeyFile,  $publicPem);

    echo "Generated and saved new RSA keypair.\n";
} else {
    $privatePem = file_get_contents($privKeyFile);
    $publicPem  = file_get_contents($pubKeyFile);
    $keyDetails = openssl_pkey_get_details(
        openssl_pkey_get_public($publicPem)
    );
    echo "Loaded existing RSA keypair.\n";
}

// === UTILITY: base64url ===

function base64url_encode(string $data): string {
    return rtrim(strtr(base64_encode($data), '+/', '-_'), '=');
}

// === BUILD JWKS ===

$rsa = $keyDetails['rsa'];
$jwk = [
    'kty' => 'RSA',
    'alg' => 'RS256',
    'use' => 'sig',
    'n'   => base64url_encode($rsa['n']),
    'e'   => base64url_encode($rsa['e']),
];
$thumbprint = base64url_encode(
    hash('sha256', json_encode([
        'e'   => $jwk['e'],
        'kty' => 'RSA',
        'n'   => $jwk['n'],
    ]), true)
);
$jwk['kid'] = $thumbprint;
$jwks = ['keys' => [$jwk]];

// === CREATE JWT ===

$now = time();
$payload = [
    'jti'          => bin2hex(random_bytes(32)),
    'iss'          => $appIssuer,
    'sub'          => $appSubject,
    'workvivo_id'  => $appWorkvivo,
    'aud'          => $appAud,
    'iat'          => $now,
    'nbf'          => $now,
    'exp'          => $now + $jwtLifetime,
    'state'        => bin2hex(random_bytes(32)),
];

$token = JWT::encode(
    $payload,
    $privatePem,
    'RS256',
    $jwk['kid']
);

// === OUTPUT ===

echo "\n=== JWT ===\n$token\n";
echo "\n=== kid ===\n{$jwk['kid']}\n";
echo "\n=== public.pem ===\n$publicPem\n";
echo "\n=== jwks ===\n" . json_encode($jwks, JSON_PRETTY_PRINT) . "\n";

// prompt to confirm JWKS is uploaded or public key
echo "\nPress [Enter] once you have updated your JWKS at:\n"
   . "https://HOST/admin/developers/apps/manage\n";
fgets(STDIN);

// === SEND REQUEST ===

$headers = [
    "Workvivo-Id: $orgId",
    "x-workvivo-jwt: $token",
    "x-workvivo-jwt-keyid: {$jwk['kid']}",
    "Accept: application/json",
    "Content-Type: application/json",
];
$postBody = json_encode(['email' => $userEmail]);

$ch = curl_init($apiUrl);
curl_setopt_array($ch, [
    CURLOPT_RETURNTRANSFER  => true,
    CURLOPT_POST            => true,
    CURLOPT_HTTPHEADER      => $headers,
    CURLOPT_POSTFIELDS      => $postBody,
    CURLOPT_SSL_VERIFYHOST  => false,
]);
$response = curl_exec($ch);
$httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
curl_close($ch);

echo "\n=== Response ($httpCode) ===\n$response\n";
